packages/babel-traverse/src/path/conversion.js

```diff
@@ -223,10 +223,14 @@ function hoistFunctionEnvironment(
     );
 
     argumentsPaths.forEach(argumentsChild => {
-      const argsRef = t.identifier(argumentsBinding);
-      argsRef.loc = argumentsChild.node.loc;
-
-      argumentsChild.replaceWith(argsRef);
+      if (
+        !argumentsChild.scope.hasOwnBinding("arguments") &&
+        !argumentsChild.scope.parent.hasOwnBinding("arguments")
```

Comment:
I'm not sure how to check this recursively 😕

Reply:
Maybe `argumentsChild.scope.hasBinding("arguments)`?

https://github.com/babel/babel/blob/5bbad8936b05a2b6739dcade6464dbc361118aa9/packages/babel-traverse/src/scope/index.js#L1036-L1048

Reply:
`.hasBinding()` checks every ancestor scope, while we should stop when we find a `function` scope.
@snitn315 You can write a `while` loop that checks if the current scope has the `arguments` binding, and then moves to `scope.parent` until it finds a function declaration/expression scope.

Another test case:
```js
var arrow = () => {
  var arguments = 2;
  return function () {
    return () => arguments;
  }
};

expect(arrow()(1, 2, 3)()).toEqual([1, 2, 3]);
```

Reply:
@nicolo-ribaudo , @mischnic  thanks, I have updated the code please review once again.